Mathlib/CategoryTheory/Sites/Discrete.lean

```diff
@@ -41,22 +41,32 @@ namespace CategoryTheory.Sheaf
 variable {C : Type*} [Category C] (J : GrothendieckTopology C) {A : Type*} [Category A]
   [HasWeakSheafify J A] {t : C} (ht : IsTerminal t)
 
-section
-variable [(constantSheaf J A).Faithful] [(constantSheaf J A).Full]
-
 /--
 A sheaf is discrete if it is a discrete object of the "underlying object" functor from the sheaf
 category to the target category.
 -/
-abbrev IsDiscrete (F : Sheaf J A) : Prop := IsIso ((constantSheafAdj J A ht).counit.app F)
+class IsDiscrete (F : Sheaf J A) : Prop :=
+  isIso_counit : IsIso ((constantSheafAdj J A ht).counit.app F)
+  faithful : (constantSheaf J A).Faithful := by infer_instance
+  full : (constantSheaf J A).Full := by infer_instance
```

Comment:
I am not super happy that a typeclass about `F` contains information that are specific to `J` and `A`. The previous design seemed better to me. Also, the notion of constant sheaf is relevant even when the functor is not fully faithful.

Reply:
That's fair. My arguments for this approach are that it just feels weird that in general, an object in the image of the constant sheaf functor is not discrete, and that this terminology follows the literature (if [nLab](https://ncatlab.org/nlab/show/discrete+object) counts as literature).

Reply:
I have some code on a different branch where I do the same cleanup for the old approach. There are some lemmas which don't require the fully faithful condition. I'll open a different PR later today or tomorrow and compare the effect on #14027 (it shouldn't matter much because there the constant sheaf functor is fully faithful).

Reply:
@joelriou do you prefer #15985?

Reply:
Yes, this seems better to me :-)